partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt

```diff
@@ -155,6 +158,150 @@ internal object PlanTyper : PlanRewriter<PlanTyper.Context>() {
         )
     }
 
+    /**
+     * Initial implementation of `EXCLUDE` schema inference. Until an RFC is finalized for `EXCLUDE`
+     * (https://github.com/partiql/partiql-spec/issues/39), this behavior is considered experimental and subject to
+     * change.
+     *
+     * There are still discussion points regarding the following edge cases
+     * - EXCLUDE on a tuple attribute that doesn't exist -- give an error/warning?
+     *   - currently no error
+     * - EXCLUDE on a tuple attribute that has duplicates -- give an error/warning? exclude one? exclude both?
+     *   - currently excludes both w/ no error
+     * - EXCLUDE on a collection index as the last step -- mark element type as optional?
+     *   - currently element type as-is
+     * - EXCLUDE on a collection index w/ remaining path steps -- mark last step's type as optional?
+     *   - currently marks last step's type as optional
+     * - EXCLUDE on a binding tuple variable (e.g. SELECT ... EXCLUDE t FROM t) -- error?
+     *   - currently a parser error
+     * - EXCLUDE on a union type -- give an error/warning? no-op? exclude on each type in union?
+     *   - currently exclude on each union type
+     * - If SELECT list includes an attribute that is excluded, we could consider giving an error in PlanTyper or
+     * some other semantic pass
+     *   - currently does not give an error
+     */
+    override fun visitRelExclude(node: Rel.Exclude, ctx: Context): Rel.Exclude {
+        val input = visitRel(node.input, ctx)
+        val exprs = node.exprs
+        val typeEnv = input.getTypeEnv()
+        val newTypeEnv = exprs.fold(typeEnv) { tEnv, expr ->
+            excludeExpr(tEnv, expr, ctx)
+        }
+        return node.copy(
+            input = input,
+            common = node.common.copy(
+                typeEnv = newTypeEnv,
+                properties = input.getProperties()
+            )
+        )
+    }
+
+    private fun excludeExpr(attrs: List<Attribute>, expr: ExcludeExpr, ctx: Context): List<Attribute> {
+        val resultAttrs = mutableListOf<Attribute>()
+        attrs.forEach { attr ->
+            val rootId = expr.root
+            if (attr.name == rootId || (expr.rootCase == Case.INSENSITIVE && attr.name.equals(expr.root, ignoreCase = true))) {
```

Comment:
I'm not sure if this would be required as part of the RFC, but If we wanted to make sure we don't come into a situation such as:
```sql
SELECT * EXCLUDE nonsense.a
FROM t
```
we could:
```kotlin
val attrsExist = attrs.find { attr -> attr.name == .. } != null
if (attrsExist.not()) { handleUnresolvedDescriptor(..) }
attrs.forEach { ->
  if (attr.name == ..) {
    ..
```

It should make sure that we don't accidentally break future queries. It's IMO better to error now even though we aren't sure what the RFC will say. Then, if we allow nonexistent bindings after the RFC, we can always remove the error? Then you can add the second test that @yliuuuu mentioned above.
> @yliuuuu A couple more test cases that can be beneficial:

LMK what you think.

Reply:
Also, it would surely help users who write:
```
SELECT a
EXCLUDE a.b
FROM <<
  { 'a': { 'b': 1, 'c': 2 } }
>> AS t
```

Since we are specifically looking in the bindings coming from the SCAN, when we look for the `EXCLUDE a.b` (and specifically looking for `a`), we are silently ignoring the fact that `a` isn't in the EXCLUDE's input bindings. So, even though the projection list can use `a`, the `EXCLUDE` is silently letting this data to pass through. The current implementation, if I'm not mistaken, will return:
```
<<
  { 'a': { 'b': 1, 'c': 2 } }
>>
```

Reply:
Since resolution of the `EXCLUDE` expressions is not defined, I think it's perfectly fine to mandate that users write `... EXCLUDE t.a.b` explicitly. Though, if we silently let them write `a.b` without notifying them that `a` isn't a binding (and therefore they think that their query will strip `b`), we'll be leaking data.

Reply:
That's fair. I had originally wanted to punt on the erroring behavior to form a consistent mental model for when to give an error in the RFC. The more I think about it though, if the exclude expr root is not in the binding tuples, then we should give some error. I'll adjust to at least give an error in this case (i.e. when the root doesn't exist)